Assets/LeapMotion/Core/Scripts/DataStructures/IIndexable.cs

```diff
@@ -0,0 +1,85 @@
+﻿using System;
+using Leap.Unity.Query;
+using System.Collections.Generic;
+
+namespace Leap.Unity {
+
+  /// <summary>
+  /// This easy-to-implement interface represents the ability to index into a collection
+  /// of elements of type T.
+  /// 
+  /// The other nice thing is that you get foreach and more complex Query support for any
+  /// IIndexable if you call Query(). The Query call is allocation-free.
+  /// 
+  /// You can also get <code>foreach</code> directly on your Indexable implementer if you
+  /// have it define a "GetEnumerator()" method that simply returns the result of the
+  /// "GetEnumerator()" extension method provided for all Indexables. (Unfortunately,
+  /// C# isn't smart enough to recognize the extension method automatically for the
+  /// purposes of <code>foreach</code>.)
+  /// </summary>
+  /// <example>
+  /// <code>foreach (var element in myIndexable.Query()) { /* ... */ }</code>
+  /// </example>
+  public interface IIndexable<T> {
+
+    T this[int idx] { get; }
+
+    int Count { get; }
+
+  }
+
+  public static class IIndexableExtensions {
+
+    public static IIndexableEnumerator<T>
+                    GetEnumerator<T>(this IIndexable<T> indexable) {
+      return new IIndexableEnumerator<T>(indexable);
+    }
+
+
+    public static QueryWrapper<T, IIndexableEnumerator<T>>
+                    Query<T>(this IIndexable<T> indexable) {
+      return new QueryWrapper<T, IIndexableEnumerator<T>>(GetEnumerator(indexable));
+    }
+
+  }
+
+  public struct IIndexableEnumerator<Element> : IQueryOp<Element> {
+
+    IIndexable<Element> indexable;
```

Comment:
Referring to an indexable by interface is fine, _unless_ the indexable is a struct, which will cause it to be boxed and allocate garbage.  This will happen with the slices (and possibly the array indexable as well?)

Reply:
Ohhh man... is this why you were thinking about having a second generic argument that was something like `K : IIndexable<T>`? I might need to give that another go in the name of remaining allocation-free

Reply:
Yeah that would be the idea, but I also tried to get it to work and couldn't get it to infer the types properly :/ Not 100% sure what to do about that.